File: cpp/questions/pathSum.cpp
#include <functional>
#include <vector>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        std::vector<long long> ps = {0};
        int ans = 0;
        std::function<void(TreeNode*)> pS = [&](TreeNode* n) -> void {
            if (n == nullptr) {
                return;
            }
            long long sps = ps.back() + n->val - targetSum;
            for (int i=0; i<ps.size(); ++i) {
                if (sps == ps[i]) {
                    ++ans;
                }
            }
            ps.push_back(sps+targetSum);
            pS(n->left);
            pS(n->right);
            ps.pop_back();
        };
        pS(root);
        return ans;
    }
};

File: python/questions/pathSum.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        ans = 0
        ps = [0]
        def pS(n: Optional[TreeNode]) -> None:
            nonlocal ans, ps
            if n is None:
                return
            sps = ps[-1] + n.val
            for p in ps:
                if sps-p == targetSum:
                    ans += 1
            ps.append(sps)
            pS(n.left)
            pS(n.right)
            ps.pop(-1)
        pS(root)
        return ans
